File: src/shared/types/document.ts
export type DocumentTarget =
  | { kind: 'workspace_file'; workspaceId: string; sourceId: string }
  | { kind: 'result'; resultId: string };

export interface DocumentSnapshot {
  target: DocumentTarget;
  revisionId: string | null;
  contentHash: string;
  format: string;
  text: string;
  editable: boolean;
  hasUnsavedDraft: boolean;
}

export interface SelectionSnapshot {
  target: DocumentTarget;
  revisionId: string | null;
  contentHash: string;
  start: number;
  end: number;
  offsetUnit: 'utf16';
  selectedTextHash: string;
}

export interface ParsedDocument {
  format: string;
  parserVersion: string;
  rawHash: string;
  extractedHash: string;
  blocks: {
    id: string;
    text: string;
    locator:
      | { kind: 'text_range'; start: number; end: number; offsetUnit: 'utf16' }
      | { kind: 'unavailable'; reason: string };
  }[];
  warnings: string[];
}

File: src-tauri/src/domain/document.rs
use crate::parser::OffsetUnit;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(
    tag = "kind",
    rename_all = "snake_case",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum DocumentTarget {
    WorkspaceFile {
        workspace_id: String,
        source_id: String,
    },
    Result {
        result_id: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DocumentSnapshot {
    pub target: DocumentTarget,
    pub revision_id: Option<String>,
    pub content_hash: String,
    pub format: String,
    pub text: String,
    pub editable: bool,
    pub has_unsaved_draft: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SelectionSnapshot {
    pub target: DocumentTarget,
    pub revision_id: Option<String>,
    pub content_hash: String,
    pub start: usize,
    pub end: usize,
    pub offset_unit: OffsetUnit,
    pub selected_text_hash: String,
}
